### ISSUE TITLE
moving from master to main mega-issue

### ISSUE THREAD

**Issue by beckermr** [conda-forge/conda-forge.github.io#1162]
to do:
 - [x] set current org-wide default to `master` for the time being so we don't break anything
 - move over non-feedstock repos/tools
   - [x] conda-smithy (https://github.com/conda-forge/conda-smithy/pull/1500)
   - [x] web services (https://github.com/conda-forge/conda-forge-webservices/pull/462,  changed heroku too)
   - [x] admin migrations (https://github.com/conda-forge/admin-migrations/pull/62)
   - [x] admin requests (https://github.com/conda-forge/admin-requests/pull/300)
   - [x] blog (https://github.com/conda-forge/blog/pull/46)
   - [x] status page (https://github.com/conda-forge/status/commit/65b48a38641d0f588e283b8be08f1ce233948023)
   - [x] docs (https://github.com/conda-forge/conda-forge.github.io/pull/1508)
   - [x] status page CI counter thingy (https://github.com/conda-forge/conda-forge-webservices/pull/461)
   - [x] cdt-builds (https://github.com/conda-forge/cdt-builds/pull/52)
   - [x] cfep
   - [x] feedstocks
   - [x] feedstock-tokens
   - [x] feedstock-outputs (https://github.com/conda-forge/feedstock-outputs/pull/23, https://github.com/conda-forge/conda-forge-webservices/pull/463, checked smithy and ci config as well)
   - [x] staged-recipes (https://github.com/conda-forge/staged-recipes/pull/16094)
   - [x] repodata-tools
   - [x] repodata-shards
   - [x] releases
   - [x] artifact validation (https://github.com/conda-forge/artifact-validation/pull/545)
   - [x] by-the-numbers (https://github.com/conda-forge/by-the-numbers/pull/16)
 - [X] figure out how to change github action workflow files and not cause rerendering issues
   - ~use ssh keys per feedstock~
   - ~have webservices add a github app token w/ workflows permissions per feedstock to its secrets before rerendering~
   - **make an option in smithy and then use an admin migration**
 - formulate plan for how to switch over feedstocks
     - [x] test github actions redirects (don't work, going to move them and then make a tag called `master`)
     - move over github actions
       - [x] automerge action (https://github.com/conda-forge/automerge-action/pull/2)
       - [x] rerender action (https://github.com/conda-forge/webservices-dispatch-action/pull/12)
     - smithy 
       - [x] PR to smithy to fully handle branch changing (https://github.com/conda-forge/conda-smithy/pull/1501)
       - [x] make a smithy release
     - [x] make webservices completely agnostic 
       - [x] remove refs to master or add main (https://github.com/conda-forge/conda-forge-webservices/pull/467)
       - [x] https://github.com/conda-forge/conda-forge-webservices/issues/468 (https://github.com/conda-forge/conda-forge-webservices/pull/469, https://github.com/conda-forge/conda-forge-webservices/pull/471)
     - [x] make admin migrations agnostic (https://github.com/conda-forge/admin-migrations/pull/63)
     - [x] make the bot agnostic (https://github.com/regro/cf-scripts/pull/1442)
     - [x] enable bot to place keys in repo secrets (https://github.com/conda-forge/conda-forge-webservices/pull/478)
     - [x] move workflows to main via admin migration (https://github.com/conda-forge/admin-migrations/pull/70)
     - [x] add new workflow files to smithy (https://github.com/conda-forge/conda-smithy/pull/1554)
     - [x] release smithy with new workflow files
     - ~[ ] turn on keys in webservices for rendering~ (not needed so didn't do it)
     - [x] make an announcement about move to main (https://github.com/conda-forge/conda-forge.github.io/pull/1609)
     - [x] move over ci-scripts repo by hand
     - [x] write admin migration to move the rest (this has to change upload_on_branch, the conda-forge.yml tooling branch name, and API calls to migrate the repo, see https://github.com/conda-forge/admin-migrations/pull/72)
     - [x] move over feedstock generation (https://github.com/conda-forge/staged-recipes/pull/18311)
     - ~[ ] close out admin migration~ (not going to do this actually in case we unarchive things)
     - [x] cleanup conda-forge/feedstocks from heroku errors (https://github.com/conda-forge/admin-migrations/pull/79)

notes:
 - github is going to use `main` 
 - they plan to provide automated tooling later this year: https://github.com/github/renaming

x-ref: https://github.com/conda-forge/conda-forge.github.io/issues/1099

**Comment by mariusvniekerk** [comment]
Given that they (gh) plan for automated tooling to do this and we have a lot of repos, I'm inclined to wait

**Comment by beckermr** [comment]
100% agreed - what they can't do for us is change our tooling so it looks for main branches instead of master

**Comment by mariusvniekerk** [comment]
i mean they could just have some magic in the background that treats those two refs as the same for everything :D

**Comment by jakirkham** [comment]
I guess the first step is to figure out where we use `master` explicitly as opposed to relying on the default branch (whatever that might be named).

**Comment by beckermr** [comment]
> I guess the first step is to figure out where we use master explicitly as opposed to relying on the default branch (whatever that might be named).

I agree. I'd rather us be proactive and make sure our tooling works. Even if github shims master to main, they could remove this later.

**Comment by beckermr** [comment]
Looks like they will be rolling out better automated support in January of 2021.

**Comment by ocefpaf** [comment]
> Given that they (gh) plan for automated tooling to do this and we have a lot of repos, I'm inclined to wait to actually rename stuff.

We are in Aug 11th and so far GH hasn't come up with a nice solution that would cover our needs. To be fair, conda-forge's needs are special and there are a lot of moving parts.

We had a quick chat about this during today's meeting and I'd like to get the ball rolling again.

@mariusvniekerk, you are usually the well of knowledge on "new stuff," did you hear of any GH change that we can use?

**Comment by beckermr** [comment]
I've started in on doing this. A lot of our admin repos and whatnot can be moved by hand.

**Comment by jakirkham** [comment]
conda-smithy seems like low hanging fruit

**Comment by beckermr** [comment]
Yep. I wanted to wait to chat with core first, but this one would be very easy indeed.

**Comment by isuruf** [comment]
Is there an example feedstock where this is done so that we can have a look before starting this?

**Comment by isuruf** [comment]
What's the difference between tooling branch and default branch? Don't see any docs on https://conda-forge.org/docs/maintainer/conda_forge_yml.html#github

**Comment by beckermr** [comment]
Yes sorry. I did not add docs yet. 

tooling branch is the branch used for github actions and references to the ci-setup feedstock. 

the default branch is the one used for setting up github when smithy is run initially etc.

**Comment by beckermr** [comment]
text for later

```
git branch -m master main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```

**Comment by beckermr** [comment]
I migrated the test feedstock here: https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/pull/390

**Comment by beckermr** [comment]
To get the readme full updated, we'd have to rerender. However, the master links should forward.

**Comment by beckermr** [comment]
@isuruf any more comments here or shall we go ahead with one of the options above?

**Comment by h-vetinari** [comment]
Saw the first feedstocks being migrated automatically - congrats for pulling into the final stretch on this!

**Comment by beckermr** [comment]
Thanks! A bit premature though. :(

When GitHub renames the branch azure sees it as a new branch and rebuilds the feedstock. 

I stopped the migration until I can put in a fix.

**Comment by jakirkham** [comment]
FWIW using [the `feedstocks` repo]( https://github.com/conda-forge/feedstocks ) and the [GitHub CLI]( https://cli.github.com ) (packaged [in conda-forge]( https://github.com/conda-forge/gh-feedstock )) wrote the following command using `bash` to update my feedstock forks to have & use the new `main` branch as default.

Note: This skips over any that haven't been migrated yet.

```bash
GH_USER=<username> git submodule foreach 'git fetch origin && (git checkout main && git pull && git push $GH_USER main && gh repo edit "$(git remote get-url $GH_USER)" --default-branch main)
```

One could also delete the `master` branches afterwards, but personally held off doing that for now in case the prior state might be needed for something. Will likely revisit after some time has passed/the migration is complete.

**Comment by beckermr** [comment]
Cool! This is running again. Should be done in a few days.

**Comment by beckermr** [comment]
OK. Everything here is done. If there are no objections, I am going to close and unpin this issue!

**Comment by h-vetinari** [comment]
Congrats on pulling this off! 🥳

**Comment by h-vetinari** [comment]
Ran into a small paper cut caused by this, not actionable but thought I'd note it if someone has an idea I've overlooked.

I wanted to open an old PR on a feedstock, but couldn't because it had been pointing against master (a branch that's now deleted), and I couldn't change the branch because the PR was closed -> Catch 22. Of course one could repush a master branch to the main feedstock (and delete it after?), but that doesn't sound like a worthwhile trade-off.

**Comment by beckermr** [comment]
Ouch. Thanks for the report. Hopefully we'll hit less of this as time goes on. Sorry for the trouble!

**Comment by jakirkham** [comment]
Usually the first step is undelete the branch in GitHub UI. Then it can usually be reopened (though have not tried that with the base branch being deleted).

The other option is open a new PR. It is possible to [fetch the commits from the old PR]( https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally ), which should minimize the restart cost in this case

**Comment by h-vetinari** [comment]
> Usually the first step is undelete the branch in GitHub UI. Then it can usually be reopened (though have not tried that with the base branch being deleted).

I did that, but the missing base branch was exactly the problem.



> The other option is open a new PR. It is possible to [fetch the commits from the old PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally), which should minimize the restart cost in this case

Indeed, and I would have done it in other cases, just that it would have been nice to redo the blas-variant [overview](https://github.com/conda-forge/numpy-feedstock/pull/196) in the same place as all the previous ones for that series.

**Comment by jakirkham** [comment]
Understood unfortunately this is out of our hands. Would suggest raising [a support issue with GitHub]( https://support.github.com/request ) and seeing what they say
